---
title: "Ancestral State Reconstruction and BAMM"
author: "Grace Musser and Erin Keenan Early"
date: "May 2, 2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
	warning = TRUE
	message = TRUE
	comment = "##"
	prompt = FALSE
	tidy = TRUE
	tidy.opts = list(blank = FALSE, width.cutoff = 75)
	fig.path = "img/"
	fig.align = "center"
```


## Ancestral State Reconstruction and BAMM


### Phylogenetics 101

To start things out lets go over some phylogenetic basics for those who are unfamiliar. We want to actually get to the coding part of this presentation, and fully understanding a plylogenetic tree and how they work is the subject of an entire class (or course), so we are presenting only a very stripped down version that will hopefully give you a core basic understanding of what it is we are doing with this code. If you have questions, please ask them, because this is definitely leaving details out! 

Phylogenetics is a system used to map relationships between organisms based upon shared ancestry in which relatedness is considered on the basis of recency of common ancestry. These relationships are visualized using phylogenetic trees, the most common two styles of which are below: 

<img src="C:/Users/keena/Documents/R/Projects/Group_Project/images/Cladogram1.jpg" width="900px"/>

These two trees show identical information, even though they look very different. 

Let's quickly go over the anatomy of a trees so you can better understand some of the terms we will be using and what exactly it is we are trying to do. 


<img src="C:/Users/keena/Documents/R/Projects/Group_Project/images/AnatomyofCaldogram.jpg" width="900px"/>

The relationships between taxa are determined by analzing characters that are shared between different taxa. So to visualize this we can look at the below image:

<img src="C:/Users/keena/Documents/R/Projects/Group_Project/images/Characteristics.jpg" width="900px"/>

So here we can look at this tree and understand some of these relationships a little bit better. For example, all of the taxa on this trees share characteristic 1, but only D, E, their common ancestor, and all of that ancestors descendants have characteristic 6. 


## So what then is Ancestral State Reconstruction? 

As paleontologists frequently we are interested in understanding the characteristics of an extinct organism to better understand the evolutionary path taken to give us our modern biota, or the characteristics of a specific organism of interest. To do this it is often helpful to highlight characters of interest and determine when in the evolutionary history certain characters developed and manifesting within a given group of organisms. Essentially what this means is we measure known characteristics of target organisms, and use those to extrapolate the relationships between all organisms of intrest. Depending on how many organisms you are interested in (called your "in group"), this can get very complicated.

<img src="C:/Users/keena/Documents/R/Projects/Group_Project/images/Cladogram2.jpg" width="900px"/>

(Citation: Graña-Miraglia L, Arreguín-Pérez C, López-Leal G, Muñoz A, Pérez-Oseguera A, Miranda-Miranda E, Cossío-Bayúgar R, Castillo-Ramírez S. 2018. Phylogenomics picks out the par excellence markers for species phylogeny in the genus Staphylococcus. PeerJ 6:e5839 https://doi.org/10.7717/peerj.5839)

So this "fan style" tree is the basic branched tree I showed you before wrapped around itself -- and we'll be making some of these with our code in just a minute. But you can see this tree is massive, and its all done based off of characters just like we looked at previously -- It's a lot to do this by hand, so programs are used instead. We are providing you with data that has been created in external programs that produce these relationships, such as PAUP* or MrBayes, which brings me to....


###Bayesian analysis

As a reminder, Bayesian analysis refers to a statistical method called Bayes' theorum, developed by Reverend Thomas Bayes in the 1700s. Bayes' theorum is used to determine the probability of an event based upon prior knowledge of conditions expected or hypothesized to be related to that event: 

<img src="C:/Users/keena/Documents/R/Projects/Group_Project/images/bayes-rule.png" width="300px"/>

This equation is read as: The probability of Event A given Event B, is equal to the probability of Event B given Event A multiplied by the probability of Event A, all divided by the probability of Event B.

So if liking chocolate is related to shoe size (Pr[liking chocolate]), we can use Bayes' theorum to determine the probability that someone will like chocolate using their shoe size (Pr[liking chocolate | shoe-size]).

Bayes' theorum is used in phylogenetics to produce the most likely plylogenetic tree for a given set of data based upon prior probabilities (i.e. character traits) -- said another way, it creates a set of models of evolutionary pathways for a known combination of characteristics seen in a target organism, and provides a probability (which we call the posterior probability, or simply "posterior"") for each node or speciation event.

We will also be looking at...


###BAMM

BAMM stands for Bayesian analysis of macroevolutionary mixtures. It is a program that is used to model speciation, extinction, and the evolution of traits on phylogenetic trees that allows us to model the rate of an evolutionary pathway through time and among groups of similar organisms (called clades).


If you want to learn more about BAMM and BAMMtools (which we will be using today) you can learn all about it here: 

http://bamm-project.org/introduction.html


So now that we have some basic background (and I hope it wasn't too confusing), let's play with some code!


#Preliminary Analyses

##Phylogenetic data matrix creation
Here, we used a combination of 13 nuclear and mitochondrial genes for avian taxa. For anyone unfamiliar, mitochondrial genes are strictly matrilineal, and nuclear genes are strictly patrilineal. 


###Models of evolution
We ran the phylogenetic analyses using Bayesian methods using MrBayes on TACC. We ran the analysis using the GTR (Generalised time reversible) model with gamma-distributed rate variation across sites. If we did not include the gamma model, it would be assumed that all sites are evolving at equal rates. The GTR model is most often used in current phylogenetic studies as it is the most complex model currently available for molecular evolution. It comprises 64 nested models with various combinations of parameters for DNA site substitution and has the highest degrees of freedom out of all previously published models of molecular evolution.

#Downstream analyses in R

##Tree extraction 

First, load the required libraries:
```{r}
library(ape)
library(BAMMtools)
library(readxl)
library(curl)
library(phytools)
```

*If you get an error asking for "quadprog", this is a dependency for phytools that you may need to download directly from https://cran.rstudio.com/bin/windows/contrib/3.5/. Just download the zip and extract it to your R win-library (or Apple equivalent) and it should work after that.


And the color palette:
```{r}
library(gameofthrones)
```


You can grab the data here:

Body mass data: (may need to be downloaded directly as excel file)
```{r}
f<-curl("https://raw.githubusercontent.com/gmusser553/ADA_Final_Project/master/ONLYGruiBodyMassData.csv")
GRUIMassData<- read.csv(f, header = TRUE, sep = ",", stringsAsFactors = TRUE)
head(GRUIMassData)
```


Posterior tree set:
```{r}
posteriors <- curl("https://raw.githubusercontent.com/gmusser553/ADA_Final_Project/master/posterior_tree_set.nex")
```

BAMM mcmcout file:
```{r}
f <- curl("https://raw.githubusercontent.com/gmusser553/ADA_Final_Project/master/mcmc_out.txt")
mcmcout <- read.csv(f, header = TRUE, sep = ",", stringsAsFactors = TRUE)
head(mcmcout)
```

Newick tree:
```{r}
my_Newick_tree<-curl("https://raw.githubusercontent.com/gmusser553/ADA_Final_Project/master/MyNewickTreefile.tre")
```


BAMM event data (may need to be downloaded and pulled from working directory)
```{r}
sevent<-curl("https://raw.githubusercontent.com/gmusser553/ADA_Final_Project/master/sevent_data.txt")
```


Extract a tree from a posterior tree set (in this case it is in nexus format). Then assign a tree from this posterior tree set to a variable. In this case we picked tree 42:
```{r}
#grui.tree<-read.nexus(file="posterior_tree_set.nex") 
grui.tree<-read.nexus(file=posteriors) 
subsettree<-grui.tree[[42]]
```

Plot the exracted tree in a fan style to check that this worked:
```{r}
plotTree(subsettree,type="fan",ftype="i", color=got(150, option = "Targaryen2"), fsize=0.3)
```

##Ancestral State Reconstruction

Here, we read in body mass data and drop names that do not match. We also take the log of the mean body mass data so that it will be easier to visalize after performing ancestral state reconstruction and plotting it on the tree.

First, download and save the following file from Github: https://github.com/gmusser553/ADA_Final_Project/blob/master/ONLYGruiBodyMassData.xlsx

```{r}
GRUIMassData <- read_excel("C:/Users/keena/Documents/R/Projects/Group_Project/ONLYGruiBodyMassData.xlsx") 
meanmass<-GRUIMassData$LogMean
names(meanmass)<-GRUIMassData$`Species Name`
xxx <- na.omit(meanmass)
v <- subsettree$tip.label %in% names(xxx)
drop <- subsettree$tip.label[v==FALSE]
d2<-drop.tip(subsettree, drop)
```


Perform ancestral state reconstruction using fastAnc:

```{r}
fit<-fastAnc(tree=d2, xxx, vars=TRUE, CI=TRUE)
fit
```

Use contMap to create a tree with the ancestral state reconstructions plotted onto it. We assign it to a variable here, but you can also use it alone to create a tree image. 

contMap takes the tree as argument x, the continuous traits to be input and from which ancestral state reconstruction can be estimated, and has several additional options that can be employed. 

Here, we decrease line width for branches since we have large taxon sampling (lwd) and increase resolution of the ancestral state reconstruction (res). fsize can be used in plot or contMap and determines font size.

```{r}
obj<-contMap(d2,xxx,plot=FALSE, res=1000, lwd=.5)
```

The setMap function allows us to change the color palette of the tree.

```{r}
obj<-setMap(obj, colors=got(128, option = "Daenerys"))
```

```{r}
plot(obj, fsize=.2)
```


We can also plot it in a fan style. Fan styles are typically used for datasets with large taxon sampling.

```{r}
obj<-setMap(obj, colors=got(256, option = "Targaryen2"))
plot(obj,type="fan",legend=0.7*max(nodeHeights(d2)),
     fsize=.5)
```

We can also plot a phenogram. A phenogram depicts taxonomic relationships based on overall similarity of character states without takiing evolutionary history or weighting/ordering of characetrs. 

```{r}
phenogram(d2,xxx,fsize=0.2,spread.costs=c(1,0))
```



##BAMM analyses and plotting in R:

```{r}
#first use BAMM in terminal-must be Newick notation
#check that tree is ultrametric and all tree lengths are greater than 0:
is.ultrametric(d2) #check if ultrametric
min(d2$edge.length) #check minimum branch length
#After running BAMM, load mcmc_out.txt or equivalent file into R to test convergence using 'coda' package:
#mcmcout <- read.csv("mcmc_out.txt", header=T)
plot(mcmcout$logLik ~ mcmcout$generation)
#next: discard burnin
burnstart <- floor(0.1 * nrow(mcmcout))
postburn <- mcmcout[burnstart:nrow(mcmcout), ]
#check the effective sample sizes of the log-likelihood and the number of shift events present in each 
#you want this to be at least 200
#effectiveSize(postburn$N_shifts)
#effectiveSize(postburn$logLik)
#tree <- read.tree("MyNewickTreefile.tre")
tree<-read.tree(my_Newick_tree)
edata <- getEventData(tree, eventdata = "https://raw.githubusercontent.com/gmusser553/ADA_Final_Project/master/sevent_data.txt", burnin=0.1)
shift_probs <- summary(edata) #create dataframe of posterior probabilities
#visualizing mean, model-averaged diversification rates at any point along every branch of a phylogenetic tree:
plot.bammdata(edata, lwd=2, legend=T)
index <- 25
e2 <- subsetEventData(edata, index = index)
plot.bammdata(e2, lwd=2, legend=TRUE)
addBAMMshifts(e2, cex=2)
```




#Phylogenetically informed correlation of BAMM rates and body mass
```{r}
D.scores <- GRUIMassData$LogMean
names(D.scores) <- GRUIMassData$`Species Name`
is.D.scores.in.edata <- names(D.scores) %in% edata$tip.label
  
D.scores.trim <- D.scores[is.D.scores.in.edata==TRUE]
  is.edata.in.D.scores <- edata$tip.label %in% names(D.scores.trim)
  keep <- edata$tip.label[is.edata.in.D.scores == TRUE]
  
  edata.subset <- subtreeBAMM(edata, tips=keep)
  
 cor.result <- traitDependentBAMM(edata.subset, traits=D.scores.trim,
                                     reps=1000, return.full = TRUE,
                                    method="s",logrates = TRUE,
                                    two.tailed = TRUE)
 
 #correlation
hist(cor.result$estimate, xlim=c(-1,1))
hist(cor.result$obs.corr, xlim=c(-1,1))
cor.result$estimate
cor.result$p.value
```

```{r}
```
